mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp

```diff
@@ -10,63 +10,146 @@
+        argMap[argKey] = token;
+        argKey.clear();
+      }
     }
+  }
+}
 
-    if (inputDim == 'n') {
-      inputDimension.push_back(batchSize);
-    } else if (inputDim == 'c') {
-      inputDimension.push_back(inputChannel);
-    } else if (inputDim == 'h') {
-      inputDimension.push_back(inputHeight);
-    } else if (inputDim == 'w') {
-      inputDimension.push_back(inputWidth);
-    }
+LogicalResult Conv2dGenerator::parseConvConfig(const char *arguments) {
+  std::map<std::string, std::string> argMap;
+  strToTokens(arguments, argMap);
+
+  auto isValid = [&argMap]() {
+    // only require tensor configs
+    static const std::vector<std::string> validKeys = {
+        "batchsize", "in_layout",  "in_type",  "in_channels",  "in_h",
+        "in_w",      "out_layout", "out_type", "out_channels", "out_h",
+        "out_w",     "fil_layout", "fil_type", "fil_w",        "fil_h"};
+    return std::all_of(
+        validKeys.cbegin(), validKeys.cend(),
+        [&argMap](const std::string &key) { return argMap.count(key) > 0; });
+  };
+
+  // Proceed only if we have a valid argMap. Otherwise leave the handle to be
+  // empty
+  if (isValid()) {
+
+    auto strToLong = [&argMap](std::string argKey) {
+      return std::stoul(argMap[argKey]);
+    };
+
+    auto strToInt = [&argMap](const std::string &key, auto &setting) {
+      if (argMap.find(key) != argMap.end()) {
+        setting = std::stoi(argMap[key]);
+      }
+    };
+
+    auto strToStr = [&argMap](const std::string &key, std::string &setting) {
+      if (argMap.find(key) != argMap.end()) {
+        setting = argMap[key];
+      }
+    };
+
+    // arch settings
+    strToStr("arch", arch);
+    strToInt("num_cu", num_cu);
+    strToInt("x2", xdlops);
+
+    // conv settings
+    operation = argMap["operation"];
+    dataTypeStr = argMap["out_type"];
+    strToInt("dilation_h", dilationHeight);
+    strToInt("dilation_w", dilationWidth);
+    strToInt("conv_stride_h", strideHeight);
+    strToInt("conv_stride_w", strideWidth);
+    strToInt("padding_h", paddingHeight);
+    strToInt("padding_w", paddingWidth);
+
+    strToStr("kernel_name", kernelName);
+
+    // MIOpen has NCHW as layout string for all three tensors
+    strToStr("fil_layout", filterLayout);
+    strToStr("in_layout", inputLayout);
+    strToStr("out_layout", outputLayout);
 
-    if (outputDim == 'n') {
-      outputDimension.push_back(batchSize);
-    } else if (outputDim == 'k') {
-      outputDimension.push_back(outputChannel);
-    } else if (outputDim == 'h') {
-      outputDimension.push_back(outputHeight);
-    } else if (outputDim == 'w') {
-      outputDimension.push_back(outputWidth);
+    // Determine tensor dimensions.
+    return parseConvDims(
+        strToLong("batchsize"), strToLong("in_channels"), strToLong("in_h"),
+        strToLong("in_w"), strToLong("out_channels"), strToLong("out_h"),
+        strToLong("out_w"), strToLong("fil_w"), strToLong("fil_h"));
```

Comment:
This function is doing too much work and can be split and updated:
 - Populate the `argMap`
 - Parsing of the `argMap`
   - Make parsing result to be some combination of return to help construct a `Conv2dGenerator`
 - Remove `parseConvDims`, make that function do orthogonal work vs this one.

Reply:
I would rather not keep argMap around, it is only needed to capture settings that would be translated into the data-members. Also I would then need to cache the batchsize, in_channels, etc.

There are 2 ways to configure the generator:

* command-line switches using parseConvConfig
* Construct generator with appropriate settings, then call parseConvDims

Reply:
I wasn't thinking clearly enough in the last comments. I think your reply help me clarify my thoughts. I meant:

> it is only needed to capture settings that would be translated into the data-members. 

I just updated the previous comments and think that the responsibility of this class may not be the storage of those data-member, but the population of a lightweight structure that hold those data-members.

 > I would rather not keep argMap around

I'm not asking the argMap to be stored, I just think that it can help splitting the parsing stage from the population stage. On the top level, you can still have the `parseConvConfig` that call the parsing + population function separately.